File: java/ecbenchmark/src/ecbenchmark/function/Ackley.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package ecbenchmark.function;

import ecbenchmark.Function;

/**
 *
 * @author jcrada
 */
public class Ackley implements Function {

    public Ackley() {
    }

    public double f(final double[] x) {
        double sumA = 0.0, sumB = 0.0;
        for (int i = 0; i < x.length; ++i) {
            sumA += x[i] * x[i];
            sumB += Math.cos(2 * Math.PI * x[i]);
        }
        return -20 * Math.exp(-0.2 * Math.sqrt((1.0 / x.length) * sumA))
                - Math.exp((1.0 / x.length) * sumB)
                + 20 + Math.E;
    }
}

File: cpp/ecbenchmark/src/function/Ackley.cpp
#include "ecbenchmark/function/Ackley.h"

#include <math.h>
namespace ecb {

    scalar Ackley::f(const std::vector<scalar>& x)  {
        scalar sumA = 0.0, sumB = 0.0;
        for (size_t  i = 0; i < x.size(); ++i) {
            sumA += x[i] * x[i];
            sumB += cos(2 * M_PI * x[i]);
        }
        return -20 * exp(-0.2 * sqrt((1.0 / x.size()) * sumA))
                - exp((1.0 / x.size()) * sumB)
                + 20 + M_E;
    }
}
